src/main/java/com/example/onlinebookstore/service/user/impl/UserServiceImpl.java

```diff
@@ -0,0 +1,45 @@
+package com.example.onlinebookstore.service.user.impl;
+
+import com.example.onlinebookstore.dto.user.UserRegistrationRequestDto;
+import com.example.onlinebookstore.dto.user.UserResponseDto;
+import com.example.onlinebookstore.exception.EntityNotFoundException;
+import com.example.onlinebookstore.exception.RegistrationException;
+import com.example.onlinebookstore.mapper.UserMapper;
+import com.example.onlinebookstore.model.Role;
+import com.example.onlinebookstore.model.User;
+import com.example.onlinebookstore.repository.role.RoleRepository;
+import com.example.onlinebookstore.repository.user.UserRepository;
+import com.example.onlinebookstore.service.user.UserService;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration.");
+        }
+        User user = new User();
+        user.setEmail(request.getEmail());
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
+        user.setFirstName(request.getFirstName());
+        user.setLastName(request.getLastName());
+        user.setShippingAddress(request.getShippingAddress());
+        Role role = roleRepository.findByName(Role.RoleName.ROLE_USER).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find role" + Role.RoleName.ROLE_USER)
+        );
```

Comment:
let's refactor it, move this logic to a separate service

        Role userRole = roleService.getRoleByRoleName(Role.RoleName.ROLE_USER);
        user.setRoles(new HashSet<>(Set.of(userRole)));